### ISSUE TITLE
Preload page index for async ParquetObjectReader

### ISSUE THREAD

**Issue by suremarc** [apache/arrow-rs#4090]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently the [`ParquetMetaData`](https://docs.rs/parquet/latest/parquet/file/metadata/struct.ParquetMetaData.html) object has optional fields for the column & offset indexes which are unpopulated at first. When the `ArrowReaderBuilder` is created using `ArrowReaderOptions::with_page_index(true)` it loads the page index at query time. However, this is potentially suboptimal as it incurs additional latency making an extra request (typically to object storage which is high-latency) for each query. 

**Describe the solution you'd like**
A new method for the `ParquetObjectReader` that toggles loading the page index at construction time, something like this:
```rust
impl ParquetObjectReader {
    pub fn preload_page_index(self, should_preload: bool) -> Self {
        self.preload_page_index = true
    }
}
```

which would trigger conditional logic in the `get_metadata` function to return metadata with the page index already loaded. 

**Describe alternatives you've considered**
A public async API for deserializing the column & offset index, similar to [`index_reader`](https://docs.rs/parquet/latest/parquet/file/page_index/index_reader/index.html) but with async support and integrated with `AsyncFileReader` to enable coalescing of multiple fetches.

**Comment by tustvold** [comment]
The design of AsyncFileReader is already written in such a way as to allow this, in particular implementations may override [get_metadata](https://docs.rs/parquet/latest/parquet/arrow/async_reader/trait.AsyncFileReader.html#tymethod.get_metadata) and return a Metadata that already has the page index loaded

**Comment by suremarc** [comment]
This is true, but there is no easy way to deserialize the page index asynchronously. Currently the easiest way I have found to do this is to fetch the relevant page index offsets, create a special implementation of [`ChunkReader`](https://docs.rs/parquet/latest/parquet/file/reader/trait.ChunkReader.html) that holds the serialized page index bytes in memory, and shove the page index data into that special structure, then use the synchronous functions in [`index_reader`](https://docs.rs/parquet/latest/parquet/file/page_index/index_reader/index.html) to deserialize it into the column & offset indexes. 

I have found that the above approach works, although extremely hacky, but I'd ask that the maintainers of this library at least consider exposing a built-in way to deserialize the page index in async code. Again, something like [`index_reader`](https://docs.rs/parquet/latest/parquet/file/page_index/index_reader/index.html) but asyncified could work if making additional changes to the `ParquetObjectReader` is not desired.

**Comment by tustvold** [comment]
:+1: I will spend some time working out a better API for this, #3851 is also related

**Comment by alamb** [comment]
PR https://github.com/apache/arrow-rs/pull/4216

**Comment by tustvold** [comment]
`label_issue.py` automatically added labels {'parquet'} from #4142
